src/line_parser.rs

```diff
@@ -0,0 +1,299 @@
+use wmidi::Note;
+
+use crate::{BeatNumber, Line, LineNote};
+
+#[derive(Debug)]
+enum State {
+    AwaitingNextToken,
+    ParsedNoteLetter { note_letter: char },
+    ParsedNoteName { note_name: String },
+}
+
+pub fn parse_line(line_str: &str) -> Line {
+    LineParser::new(line_str).parse()
+}
+
+struct LineParser {
+    chars: Vec<char>,
+    next_index: usize,
+    state: State,
+    currently_sustained_note: Option<(Note, BeatNumber)>,
+    current_beat_number: BeatNumber,
+}
+
+impl LineParser {
+    pub fn new(line_str: &str) -> Self {
+        Self {
+            chars: line_str.chars().collect(),
+            next_index: 0,
+            state: State::AwaitingNextToken,
+            currently_sustained_note: None,
+            current_beat_number: BeatNumber { sixteenth_note: 0 },
+        }
+    }
+
+    pub fn parse(&mut self) -> Line {
+        let mut line_notes: Vec<LineNote> = vec![];
+
+        while self.next_index < self.chars.len() {
+            let next_char = self.chars[self.next_index];
+            match (next_char, &self.state) {
+                (' ', State::AwaitingNextToken) => {
+                    self.forward(1);
```

Comment:
Since all of these cases are moving forward one char at a time, could just express this as a `for` loop over `self.chars`